### PR TITLE
add rag benchmarking notebook example

### PR DESCRIPTION
**Pull request by skapadia3214**

### CODE REVIEW

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 1]
```diff
@@ -0,0 +1,731 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "\n",
+    "# Benchmarking a RAG Pipeline using Llama3 and Langchain\n",
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:8]:
```suggestion
    "# Benchmarking a RAG Pipeline with Llama3 and LangChain\n",
```

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 2]
```diff
@@ -0,0 +1,731 @@
...
+    "\n",
+    "Retrieval Augmented Generation (RAG) pipelines are one of the most popular use cases of Generative AI and LLMs.  Evaluating these pipelines are crucial for understanding their performance, limitations, and potential for improvement. \n",
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:10]:
```suggestion
    "Retrieval Augmented Generation (RAG) pipelines are one of the most popular use cases of Generative AI and Large Language Models (LLMs).  Evaluating these pipelines is crucial for understanding their performance, limitations, and potential for improvement. \n",
```

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 3]
```diff
@@ -0,0 +1,731 @@
...
+    "\n",
+    "This notebook demonstrates how to benchmark a RAG pipeline using Llama3-8B as the retrieval model and Llama3-70B as both the dataset generator and the evaluation judge.\n",
+    "\n",
+    "Here is what we will do:\n",
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:14]:
```suggestion
    "This tutorial will walk you through how to:\n",
```

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 4]
```diff
@@ -0,0 +1,731 @@
...
+    "\n",
+    "1. Set up the environment and dependencies.\n",
+    "2. Load and index the data into the vectorstore.\n",
+    "3. Set up the RAG pipeline.\n",
+    "4. Generate evaluation dataset.\n",
+    "5. Evaluate the RAG pipeline.\n",
+    "6. Display and save the evaluation metrics.\n",
+    "\n",
+    "We use Llama3-8B and Llama3-70B through [Groq](https://groq.com) - you can sign up there to get a free trial API key."
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:23]:
```suggestion
    "For the fastest LLM inference speed in the world, we use Llama3-8B and Llama3-70B powered by Groq. You can create a developer account for free at https://console.groq.com/ and generate a free API key to follow this tutorial!"
```

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 5]
```diff
@@ -0,0 +1,731 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 1. Set up the Environment and Dependencies\n",
+    "\n",
+    "In this section, we install and import the necessary libraries required for our benchmarking task."
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:32]:
```suggestion
    "In this section, we install and import the necessary libraries required for our benchmarking task."
    
    ### Install and Import Libraries
```

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 6]
```diff
@@ -0,0 +1,731 @@
...
+   "source": [
+    "import nest_asyncio\n",
+    "nest_asyncio.apply()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Setup Environment Variables\n",
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:65]:
```suggestion
    "### Set Up Environment Variables\n",
```

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 7]
```diff
@@ -0,0 +1,731 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Instantiate LLM and Embeddings Model\n",
+    "\n",
+    "We then instantiate the `ChatGroq` class to use Llama3 through Groq and `HuggingFaceEmbeddings` to use the embeddings model.\n",
+    "\n",
+    "NOTE: The `BBAI/bge-small-en-v1.5` HuggingFace embeddings model will download the model locally on your computer. The model is ~135MB.\n",
+    "\n",
+    "You can swap out any of the Chat and Embeddings model with any of Langchain's [Chat Model integrations](https://python.langchain.com/v0.2/docs/integrations/chat/) and [Embeddings Model integrations](https://python.langchain.com/v0.2/docs/integrations/text_embedding/)."
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:92]:
```suggestion
    "You can swap out any of the Chat and Embeddings models with any of Langchain's [Chat Model integrations](https://python.langchain.com/v0.2/docs/integrations/chat/) and [Embeddings Model integrations](https://python.langchain.com/v0.2/docs/integrations/text_embedding/)."
```

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 8]
```diff
@@ -0,0 +1,731 @@
...
+    "rag_llm = ChatGroq(model=\"llama3-8b-8192\") # Used for RAG\n",
+    "qa_llm = ChatGroq(model=\"llama3-70b-8192\", temperature=0.1) # Used to create eval dataset\n",
+    "benchmark_llm = ChatGroq(model=\"llama3-70b-8192\") # Used to evaluate (Judge)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 2. Load and index the data into the vectorstore.\n",
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:127]:
```suggestion
    "## 2. Load and Index Data into the Vectorstore\n",
```

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 9]
```diff
@@ -0,0 +1,731 @@
...
+   ],
+   "source": [
+    "await retriever.ainvoke(\"What did paul graham do growing up?\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 3. Set up the RAG pipeline\n",
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:250]:
```suggestion
    "## 3. Set up the RAG Pipeline\n",
```

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 10]
```diff
@@ -0,0 +1,731 @@
...
+   ],
+   "source": [
+    "await rag_chain.ainvoke(\"What did paul graham do growing up?\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 4. Generate evaluation dataset\n",
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:330]:
```suggestion
    "## 4. Generate Evaluation Dataset\n",
```

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 11]
```diff
@@ -0,0 +1,731 @@
...
+    "\n",
+    "To evaluate our RAG pipeline, we need to generate a dataset with questions that are relevant to our data.\n",
+    "\n",
+    "Groq supports providing output json by using the `json_mode`. We will use this to generate formatted questions for each document from our data.\n",
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:334]:
```suggestion
    "Groq supports providing output JSON by using the `json_mode` feature. We will use `json_mode` to generate formatted questions for each document from our data.\n",
```

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 12]
```diff
@@ -0,0 +1,731 @@
...
+    "\n",
+    "We create the  `qa_chain` that takes in a chunk of text and returns 3 questions in json format."
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:336]:
```suggestion
    "We create the `qa_chain` that takes in a chunk of text and returns 3 questions in JSON format."
```

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 13]
```diff
@@ -0,0 +1,731 @@
...
+    "print(f\"From document: \\n{texts[0]}\\n\")\n",
+    "print(f\"Questions generated:\")\n",
+    "for i, q in enumerate(questions[0].values(), 1): print(f'{i}: {q}')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 5. Evaluate the RAG pipeline\n",
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:439]:
```suggestion
    "## 5. Evaluate the RAG Pipeline\n",
```

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 14]
```diff
@@ -0,0 +1,731 @@
...
+    "\n",
+    "Now we create a pipeline to evaluate our RAG pipeline on our generate set of questions. To get an evaluation we need to provide the question, document that the question was generated from and an answer from the RAG pipeline. We pass all this into our `benchmark_llm` and use `json_mode` to return a score and an evalation of the answer."
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:441]:
```suggestion
    "Now let's create a pipeline to evaluate our RAG pipeline on our generated set of questions. To get an evaluation, we need to provide the question, document that the question was generated from, and an answer from the RAG pipeline. We pass all this into our `benchmark_llm` and use `json_mode` to return a score as well as an evaluation of the answer."
```

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 15]
```diff
@@ -0,0 +1,731 @@
...
+    "    | benchmark_llm.with_structured_output(schema=EvalResponse, method='json_mode') # Parse response according to EvalResponse object\n",
+    ")\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Lets try one question, context, answer pair in our newly created `benchmark_chain`."
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:498]:
```suggestion
    "Lets test a question and context/answer pair in our newly created `benchmark_chain`."
```

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 16]
```diff
@@ -0,0 +1,731 @@
...
+   ],
+   "source": [
+    "evaluations = await evaluate(questions[:5], texts[:5]) # Remove the `:5` to evaluate all the questions on all your data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## 6. Display and save the evaluation metrics\n",
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:623]:
```suggestion
    "## 6. Display and Save the Evaluation Metrics\n",
```

File: benchmarking-rag-langchain/benchmarking_rag.ipynb [hunk 17]
```diff
@@ -0,0 +1,731 @@
...
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.12.3"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 2
+}
```
Comment (hozen-groq) [inline benchmarking-rag-langchain/benchmarking_rag.ipynb:731]: @skapadia3214 Please also include a short, few-sentence conclusion that recaps what we learned in this tutorial

Comment (skapadia3214) [reply benchmarking-rag-langchain/benchmarking_rag.ipynb:751]: Added a conclusion in [this commit](https://github.com/groq/groq-api-cookbook/pull/20/commits/65ee5eb3ef2fc2e65cc16d150d1ddbba24a85e3d)

**Pull request merged**